libexec/job/slurm/monitor.sh

```diff
@@ -75,75 +82,51 @@ read -r -d '' template <<'TEMPLATE' || true
+        --arg reason "${PARSE_RESULT[reason]}" \
+        --arg estimated_start_time "${PARSE_RESULT[estimated_start_time]}" \
+        --arg estimated_end_time "${PARSE_RESULT[estimated_end_time]}" \
+        --arg start_time "${PARSE_RESULT[start_time]}" \
+        --arg end_time "${PARSE_RESULT[end_time]}" \
+        --arg stdout_path "${PARSE_RESULT[stdout_path]}" \
+        --arg stderr_path "${PARSE_RESULT[stderr_path]}" \
+        "$template"
+}
 
-if [[ "$exit_status" -eq 0 ]]; then
-  state=$(                parse_scontrol_state  "$raw_control")
-  scheduler_state=$(      parse_scontrol_scheduler_state "$raw_control")
-  reason=$(               parse_scontrol_reason "$raw_control")
-  start_time=$(           parse_scontrol_start_time "$raw_control" "$state")
-  end_time=$(             parse_scontrol_end_time   "$raw_control" "$state")
-  estimated_start_time=$( parse_scontrol_estimated_start_time "$raw_control" "$state")
-  estimated_end_time=$(   parse_scontrol_estimated_end_time   "$raw_control" "$state")
-  stdout_path=$(          parse_scontrol_stdout "$raw_control")
-  stderr_path=$(          parse_scontrol_stderr "$raw_control")
-elif [[ "$raw_control" == "slurm_load_jobs error: Invalid job id specified" ]]; then
-  # Fallback to sacct if scontrol does not recognise the ID
-  raw_acct=$(sacct --noheader --parsable --jobs "$1" --format State,Reason,START,END,AllocTRES)
-  exit_status="$?"
-  acct=$(echo "$raw_acct" | head -n1)
-  cat <<EOF >&2
+main() {
+    declare -A PARSE_RESULT
+    local exit_status
+    local output
 
-sacct:
-$raw_acct
-EOF
+    assert_progs jq scontrol sacct
 
-  # Unset all the fields if the job is UNKNOWN
-  if [[ "$exit_status" -eq 0 ]] && [ -z "$acct" ]; then
-    state="UNKNOWN"
-    scheduler_state=""
-    start_time=''
-    end_time=''
-    reason=''
+    output=$(run_scontrol "$1" | tee >(log_command "scontrol" 1>&2))
+    exit_status=$?
 
-    # Extract the output from sacct
-  elif [[ "$exit_status" -eq 0 ]]; then
-    state=$(                parse_sacct_state  "$acct")
-    scheduler_state=$(      parse_sacct_scheduler_state "$acct")
-    reason=$(               parse_sacct_reason "$acct")
-    start_time=$(           parse_sacct_start_time "$acct" "$state")
-    end_time=$(             parse_sacct_end_time   "$acct" "$state")
-    estimated_start_time=$( parse_sacct_estimated_start_time "$acct" "$state")
-    estimated_end_time=$(   parse_sacct_estimated_end_time   "$acct" "$state")
+    if [[ $exit_status -eq 0 ]] ; then
+        output="$(echo "$output" | head -n 1 | tr ' ' '\n')"
+        source_parsers "scontrol"
+        parse_task <<< "${output}"
+    elif [[ "${output}" == "slurm_load_jobs error: Invalid job id specified" ]] ; then
+        output=$(run_sacct "$1" | tee >(log_command "sacct" 1>&2))
+        exit_status=$?
+        output=$(echo "$output" | head -n1)
+        if [[ $exit_status -eq 0 ]] && [ -z "$output" ]; then
+            PARSE_RESULT[state]="UNKNOWN"
```

Comment:
How so?  In this case, the only thing that we know about the job is that it is unknown.  That is, its state is `UNKNOWN`.

We need to create a JSON template to represent that.  The `generate_template` uses `PARSE_RESULT` to create that template.  It seems to me that this is using `PARSE_RESULT` exactly how it should be used.

What do you think the purpose of `PARSE_RESULT` is?

Reply:
The problem is `PARSE_RESULT` is a global, so it makes it difficult to determine its state at any particular point in time.

A way out of this problem, is to be rather strict about when it is modified. From its name, I would expect it only to be modified within the parsing library.

---

In regards to the `generate_template`, it can probably get away with using a local variable. So I don't see that as a particularly big blocker.